File: Scripts/Networking/net-check_rs/network_checks.py
#!/usr/bin/env python3

"""
Network Check Functions for Network Defense Monitor

Contains functions to verify network interface status, IP configuration,
external IP, and DNS settings.
"""

import logging
import re
import socket
from enum import Enum, auto
from ipaddress import ip_address, ip_network, AddressValueError, NetworkValueError, IPv4Address, IPv6Address
from typing import List, Optional, Tuple, Set

import psutil
import requests

# Import Config from main script (adjust path if necessary)
# This assumes network_checks.py is in the same directory as net_defense_monitor.py
from net_defense_monitor import Config

# --- Result Enums (Matching Rust Enums) ---

class IpNetworkMatchStatus(Enum):
    NOT_CHECKED = auto()        # No expected network was specified.
    MATCH = auto()              # Found an IP matching the expected network.
    MISMATCH = auto()           # Found IPs, but none matched the expected network.
    NO_IPS = auto()             # Expected a network, but the interface had no IPs.
    INTERFACE_HAS_NO_IPS = auto() # No expected network, and interface has no IPs (warning).

class ExternalIpStatus(Enum):
    DISABLED = auto()             # Check was not enabled.
    CHECK_FAILED = auto()         # The check itself failed (e.g., network error).
    CHECKED_OK = auto()           # Check successful, no expected IPs configured.
    EXPECTED_MATCH = auto()       # Reported IP matches one of the expected IPs.
    EXPECTED_MISMATCH = auto()    # Reported IP does not match expected IPs.

class DnsMatchStatus(Enum):
    NOT_CHECKED = auto()        # No expected DNS servers were specified.
    READ_ERROR = auto()         # Failed to read or parse resolv.conf
    MATCH = auto()              # Found servers exactly match expected servers.
    MISMATCH = auto()           # Found servers do not match expected servers.
    NO_SERVERS_FOUND = auto()   # resolv.conf parsed, but no nameserver lines found.

# --- Result Dataclasses ---
# Re-define here for clarity, matching main script structure but adding details

@dataclass
class InterfaceCheckResult:
    interface_name: str
    interface_found: bool = False
    is_up: Optional[bool] = None
    is_running: Optional[bool] = None # psutil doesn't directly map to "is_running", use is_up for now
    is_loopback: Optional[bool] = None
    ip_addresses: List[ip_network] = field(default_factory=list)
    mac_address: Optional[str] = None
    ip_network_match_status: IpNetworkMatchStatus = IpNetworkMatchStatus.NOT_CHECKED
    expected_ip_network: Optional[ip_network] = None # Store what was expected
    external_ip_status: ExternalIpStatus = ExternalIpStatus.DISABLED
    reported_external_ip: Optional[IPv4Address | IPv6Address] = None
    expected_external_ips: Optional[List[IPv4Address | IPv6Address]] = None
    findings: List[str] = field(default_factory=list)

@dataclass
class DnsCheckResult:
    servers_expected: List[IPv4Address | IPv6Address] = field(default_factory=list)
    servers_found: List[IPv4Address | IPv6Address] = field(default_factory=list)
    match_status: DnsMatchStatus = DnsMatchStatus.NOT_CHECKED
    resolv_conf_path: str = "/etc/resolv.conf"
    findings: List[str] = field(default_factory=list)

# --- Helper Functions ---

def _get_interface_stats(if_name: str) -> Optional[psutil._common.snicstats]:
    """Gets status flags for a given interface name using psutil."""
    try:
        stats = psutil.net_if_stats()
        return stats.get(if_name)
    except Exception as e:
        logging.error(f"Error getting interface stats for {if_name}: {e}")
        return None

def _get_interface_addrs(if_name: str) -> Tuple[List[ip_network], Optional[str]]:
    """Gets IP addresses and MAC address for a given interface name using psutil."""
    ips = []
    mac = None
    try:
        addrs = psutil.net_if_addrs()
        if if_name in addrs:
            for snicaddr in addrs[if_name]:
                if snicaddr.family == psutil.AF_LINK:
                    mac = snicaddr.address
                elif snicaddr.family in (socket.AF_INET, socket.AF_INET6):
                    try:
                        # Construct ip_network, requires netmask
                        if snicaddr.netmask:
                            ip_str = f"{snicaddr.address}/{snicaddr.netmask}"
                            ips.append(ip_network(ip_str, strict=False))
                        else:
                            # Handle cases without netmask (e.g., point-to-point)
                            ips.append(ip_network(snicaddr.address, strict=False))
                            logging.debug(f"Interface {if_name} address {snicaddr.address} has no netmask.")
                    except (AddressValueError, NetworkValueError, TypeError) as e:
                        logging.warning(f"Could not parse IP/Network {snicaddr.address}/{snicaddr.netmask} for {if_name}: {e}")
    except Exception as e:
        logging.error(f"Error getting interface addresses for {if_name}: {e}")

    return ips, mac

def _perform_external_ip_check(config: Config) -> Tuple[ExternalIpStatus, Optional[IPv4Address | IPv6Address], Optional[str]]:
    """Performs the external IP check using the configured URL."""
    url = config.external_ip_check_url or "https://ifconfig.me/ip"
    logging.debug(f"Querying external IP check service: {url}")
    finding = None
    reported_ip = None

    try:
        response = requests.get(url, timeout=10)
        response.raise_for_status() # Raise HTTPError for bad responses (4xx or 5xx)

        # Try parsing as plain text first
        ip_str = response.text.strip()
        try:
            reported_ip = ip_address(ip_str)
        except ValueError:
            # If plain text fails, try common JSON format {"ip": "..."}
            try:
                json_response = response.json()
                if isinstance(json_response, dict) and 'ip' in json_response:
                     ip_str = json_response['ip'].strip()
                     reported_ip = ip_address(ip_str)
                else:
                     raise ValueError("JSON response lacks 'ip' key")
            except (requests.exceptions.JSONDecodeError, ValueError) as json_e:
                finding = f"Failed to parse response from {url} as IP or JSON: {json_e}. Content: '{response.text[:100]}...'"
                logging.error(finding)
                return ExternalIpStatus.CHECK_FAILED, None, finding

        # Successfully parsed IP
        logging.info(f"External IP check service reported IP: {reported_ip}")
        expected_ips = config.expected_external_ips or []

        if not expected_ips:
            return ExternalIpStatus.CHECKED_OK, reported_ip, None
        elif reported_ip in expected_ips:
            logging.info("Reported external IP matches one of the expected IPs.")
            return ExternalIpStatus.EXPECTED_MATCH, reported_ip, None
        else:
            finding = f"Reported external IP {reported_ip} does NOT match any expected IPs: {expected_ips}"
            logging.error(finding)
            return ExternalIpStatus.EXPECTED_MISMATCH, reported_ip, finding

    except requests.exceptions.RequestException as e:
        finding = f"External IP check to {url} failed: {e}"
        logging.error(finding)
        return ExternalIpStatus.CHECK_FAILED, None, finding
    except Exception as e:
        finding = f"Unexpected error during external IP check: {e}"
        logging.error(finding, exc_info=True)
        return ExternalIpStatus.CHECK_FAILED, None, finding

# --- Main Check Functions ---

def verify_network_interfaces(config: Config) -> InterfaceCheckResult:
    """Verifies network interfaces using psutil.

    Checks the VPN interface status and IP configuration.
    Optionally performs an external IP check.
    Collects findings instead of raising exceptions.
    """
    if_name = config.vpn_interface_name
    logging.info(f"Verifying network interface: {if_name}...")

    result = InterfaceCheckResult(interface_name=if_name,
                                expected_ip_network=config.expected_vpn_ip_network,
                                expected_external_ips=config.expected_external_ips)

    stats = _get_interface_stats(if_name)
    ips, mac = _get_interface_addrs(if_name)

    if stats is not None:
        result.interface_found = True
        result.is_up = stats.isup
        # psutil doesn't have a direct 'is_running' like Rust's pnet.
        # isup is the closest, indicating if admin enabled and link detected (usually).
        # We can use is_up for the 'running' check logic from Rust.
        result.is_running = stats.isup
        # Check flags attribute for loopback (requires psutil >= 5.8.0 maybe? check docs)
        # result.is_loopback = 'LOOPBACK' in stats.flags # Example, check actual flags
        # psutil doesn't easily expose point-to-point or loopback flags directly in stats.
        # We might need more specific checks if those flags are critical.
        result.mac_address = mac
        result.ip_addresses = ips

        logging.info(f"Found interface: {if_name}")
        logging.debug(f"Interface stats: {stats}")
        logging.debug(f"Interface IPs: {ips}")
        logging.debug(f"Interface MAC: {mac}")

        if not result.is_up:
            finding = f"Interface {if_name} is down."
            result.findings.append(finding)
            logging.error(finding)
            # If down, skip IP and external checks
            result.ip_network_match_status = IpNetworkMatchStatus.NOT_CHECKED # Or a new status? InterfaceDown?
            result.external_ip_status = ExternalIpStatus.DISABLED # Or InterfaceDown?
            return result
        else:
            logging.info(f"Interface {if_name} is up.")
            # Check IP network matching
            expected_network = config.expected_vpn_ip_network
            if expected_network:
                found_match = False
                if not ips:
                    finding = f"Interface {if_name} has no IP addresses, but expected network {expected_network} was configured."
                    result.findings.append(finding)
                    logging.error(finding)
                    result.ip_network_match_status = IpNetworkMatchStatus.NO_IPS
                else:
                    for ip_net in ips:
                        # Check if the interface IP (ip_net.ip) is within the expected network
                        if ip_net.ip in expected_network:
                            logging.info(f"Found expected IP {ip_net.ip} (in {ip_net}) within network {expected_network} on interface {if_name}.")
                            found_match = True
                            break
                    if found_match:
                        result.ip_network_match_status = IpNetworkMatchStatus.MATCH
                    else:
                        finding = f"No IP address on interface {if_name} belongs to the expected network {expected_network}. Found IPs: {ips}"
                        result.findings.append(finding)
                        logging.error(finding)
                        result.ip_network_match_status = IpNetworkMatchStatus.MISMATCH
            else:
                # No expected network, just check if any IP exists
                if not ips:
                    finding = f"Warning: Interface {if_name} has no assigned IP addresses. (No expected network was configured)."
                    result.findings.append(finding)
                    logging.warning(finding)
                    result.ip_network_match_status = IpNetworkMatchStatus.INTERFACE_HAS_NO_IPS
                else:
                    logging.info(f"IP addresses found for {if_name}: {ips}. (No specific network expected).")
                    result.ip_network_match_status = IpNetworkMatchStatus.NOT_CHECKED

            # Perform external IP check if enabled
            if config.check_external_ip:
                logging.info("--- Performing External IP Check ---")
                logging.warning("OPSEC WARNING: External IP check enabled. This sends traffic outside the VPN potentially, revealing real IP if VPN fails.")
                status, reported_ip, finding = _perform_external_ip_check(config)
                result.external_ip_status = status
                result.reported_external_ip = reported_ip
                if finding:
                    result.findings.append(finding)
            else:
                logging.info("--- Skipping External IP Check (disabled) ---")
                result.external_ip_status = ExternalIpStatus.DISABLED

    else:
        finding = f"Configured interface {if_name} not found or failed to get stats."
        result.findings.append(finding)
        logging.error(finding)
        result.interface_found = False

    return result

def verify_dns(config: Config) -> DnsCheckResult:
    """Verifies DNS configuration by reading /etc/resolv.conf.

    Compares found nameservers against the expected list in the config.
    """
    logging.info("Verifying DNS configuration...")
    resolv_conf_path = "/etc/resolv.conf"
    expected_servers = config.expected_dns_servers or []

    result = DnsCheckResult(servers_expected=expected_servers,
                              resolv_conf_path=resolv_conf_path)

    if not expected_servers:
        logging.warning("No expected DNS servers configured. Skipping DNS check.")
        result.match_status = DnsMatchStatus.NOT_CHECKED
        return result

    logging.debug(f"Reading DNS configuration from {resolv_conf_path}")
    logging.debug(f"Expected DNS servers: {expected_servers}")

    found_servers_set: Set[IPv4Address | IPv6Address] = set()
    try:
        with open(resolv_conf_path, 'r') as f:
            # Regex to find lines starting with "nameserver" followed by an IP address.
            # Allows for potential comments after the IP
            nameserver_re = re.compile(r"^\s*nameserver\s+([^\s#]+)")
            for line_num, line in enumerate(f):
                line = line.strip()
                if not line or line.startswith('#'):
                    continue

                match = nameserver_re.match(line)
                if match:
                    server_str = match.group(1)
                    try:
                        ip = ip_address(server_str)
                        found_servers_set.add(ip)
                    except ValueError as e:
                        finding = f"Warning: Failed to parse potential nameserver IP '{server_str}' from {resolv_conf_path} (line {line_num+1}): {e}. Skipping."
                        result.findings.append(finding)
                        logging.warning(finding)

    except FileNotFoundError:
        finding = f"DNS configuration file not found: {resolv_conf_path}"
        result.findings.append(finding)
        logging.error(finding)
        result.match_status = DnsMatchStatus.READ_ERROR
        return result
    except IOError as e:
        finding = f"Failed to read {resolv_conf_path}: {e}"
        result.findings.append(finding)
        logging.error(finding)
        result.match_status = DnsMatchStatus.READ_ERROR
        return result
    except Exception as e:
        finding = f"Unexpected error reading or parsing {resolv_conf_path}: {e}"
        result.findings.append(finding)
        logging.error(finding, exc_info=True)
        result.match_status = DnsMatchStatus.READ_ERROR
        return result

    result.servers_found = sorted(list(found_servers_set)) # Store sorted list

    if not found_servers_set:
        finding = f"No valid 'nameserver' entries found in {resolv_conf_path}"
        result.findings.append(finding)
        logging.error(finding)
        result.match_status = DnsMatchStatus.NO_SERVERS_FOUND
        return result

    logging.info(f"DNS servers found in {resolv_conf_path}: {result.servers_found}")

    # Compare found set with expected set (order-insensitive)
    expected_servers_set = set(expected_servers)

    if found_servers_set == expected_servers_set:
        logging.info("Actual DNS servers match expected configuration.")
        result.match_status = DnsMatchStatus.MATCH
    else:
        missing = sorted(list(expected_servers_set - found_servers_set))
        unexpected = sorted(list(found_servers_set - expected_servers_set))
        error_msg = "DNS server configuration mismatch."
        if missing:
            error_msg += f" Missing expected servers: {missing}."
        if unexpected:
            error_msg += f" Found unexpected servers: {unexpected}."

        result.findings.append(error_msg)
        logging.error(error_msg)
        result.match_status = DnsMatchStatus.MISMATCH

    return result 

File: Scripts/Networking/net-check_rs/src/network_checks.rs
use crate::Config; // Use Config from the main module (since config.rs is deleted)
use log::{info, warn, debug, error}; // Logging
use pnet::datalink; // Network interface listing
use std::fs; // Filesystem access for resolv.conf
use std::collections::HashSet; // For comparing DNS servers
use regex::Regex; // For parsing resolv.conf
use std::net::IpAddr;
use std::net::IpAddr::V4; // Import specific variants if needed for clarity
use std::net::IpAddr::V6;
use ipnetwork::IpNetwork; // For IpNetwork type used in results
use reqwest::blocking::Client; // Added for external IP check
use serde::Deserialize; // Added for parsing JSON response from IP check service

/// Represents the outcome of matching an IP against an expected network.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum IpNetworkMatchStatus {
    NotChecked,         // No expected network was specified.
    Match(IpNetwork),   // Found an IP matching the expected network.
    Mismatch(IpNetwork),// Found IPs, but none matched the expected network.
    NoIPs(IpNetwork),   // Expected a network, but the interface had no IPs.
    InterfaceHasNoIPs,  // No expected network, and interface has no IPs (warning).
}

/// Represents the outcome of the optional external IP check.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ExternalIpStatus {
    Disabled,                      // Check was not enabled.
    CheckFailed(String),           // The check itself failed (e.g., network error).
    CheckedOk(IpAddr),             // Check successful, no expected IPs configured.
    ExpectedMatch(IpAddr),         // Reported IP matches one of the expected IPs.
    ExpectedMismatch(IpAddr, Vec<IpAddr>), // Reported IP does not match expected IPs.
}

/// Holds the detailed results of the network interface verification.
#[derive(Debug, Clone)]
pub struct InterfaceCheckResult {
    pub interface_name: String,
    pub interface_found: bool,
    pub is_up: Option<bool>, // None if not found
    pub is_running: Option<bool>, // None if not found
    pub is_loopback: Option<bool>, // None if not found
    pub is_point_to_point: Option<bool>, // None if not found
    pub ip_addresses: Vec<IpNetwork>, // IPs found on the interface
    pub mac_address: Option<String>, // MAC address if available
    pub ip_network_match_status: IpNetworkMatchStatus,
    pub external_ip_status: ExternalIpStatus,
    /// Collected warnings or errors found during the check.
    pub findings: Vec<String>,
}

/// Represents the outcome of matching DNS servers.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum DnsMatchStatus {
    NotChecked,     // No expected DNS servers were specified.
    ReadError(String), // Failed to read or parse resolv.conf
    Match,          // Found servers exactly match expected servers.
    Mismatch,       // Found servers do not match expected servers.
    NoServersFound, // resolv.conf parsed, but no nameserver lines found.
}

/// Holds the detailed results of the DNS verification.
#[derive(Debug, Clone)]
pub struct DnsCheckResult {
    pub servers_expected: Vec<IpAddr>,
    pub servers_found: Vec<IpAddr>,
    pub match_status: DnsMatchStatus,
    /// Collected warnings or errors found during the check.
    pub findings: Vec<String>,
}

/// Helper struct to parse simple JSON IP response (e.g., {"ip": "1.2.3.4"})
#[derive(Deserialize, Debug)]
struct IpCheckResponse {
    ip: String,
}

/// Verifies network interfaces, checking the VPN interface status and IP configuration.
/// Optionally performs an external IP check.
/// Collects findings instead of returning early errors.
///
/// # Arguments
/// * `config` - The application configuration (hardcoded defaults).
/// # Returns
/// An `InterfaceCheckResult` struct containing detailed results and findings.
pub fn verify_network_interfaces(config: &Config) -> InterfaceCheckResult {
    info!("Verifying network interfaces...");
    debug!("Looking for VPN interface: {}", config.vpn_interface_name);

    let mut result = InterfaceCheckResult {
        interface_name: config.vpn_interface_name.clone(),
        interface_found: false,
        is_up: None,
        is_running: None,
        is_loopback: None,
        is_point_to_point: None,
        ip_addresses: Vec::new(),
        mac_address: None,
        ip_network_match_status: IpNetworkMatchStatus::NotChecked, // Default
        external_ip_status: ExternalIpStatus::Disabled, // Default
        findings: Vec::new(),
    };

    let interfaces = datalink::interfaces();
    let vpn_iface = interfaces.iter().find(|iface| iface.name == config.vpn_interface_name);

    match vpn_iface {
        Some(iface) => {
            result.interface_found = true;
            result.is_up = Some(iface.is_up());
            result.is_running = Some(iface.is_running());
            result.is_loopback = Some(iface.is_loopback());
            result.is_point_to_point = Some(iface.is_point_to_point());
            result.ip_addresses = iface.ips.clone();
            result.mac_address = iface.mac.map(|mac| mac.to_string());

            info!("Found VPN interface: {}", iface.name);
            debug!("Interface details: {:?}", iface);

            if !iface.is_up() {
                result.findings.push(format!(
                    "VPN interface {} is down.", config.vpn_interface_name
                ));
            }
            if !iface.is_running() {
                 // This might be acceptable briefly, but flag it.
                result.findings.push(format!(
                    "Warning: VPN interface {} is up but not running (may lack carrier or be misconfigured).",
                    config.vpn_interface_name
                ));
            }
            if iface.is_loopback() {
                result.findings.push(format!(
                    "Warning: VPN interface {} is a loopback interface.", config.vpn_interface_name
                ));
            }
            if iface.is_point_to_point() {
                info!("VPN interface {} is a point-to-point interface.", config.vpn_interface_name);
            }

            if iface.is_up() { // Only proceed with IP checks if interface is up
                info!("VPN interface {} is up.", config.vpn_interface_name);

                // Check IP addresses associated with the interface against the expected network.
                match &config.expected_vpn_ip_network {
                    Some(expected_network) => {
                        let mut found_matching_ip = false;
                        if iface.ips.is_empty() {
                            result.findings.push(format!(
                                "VPN interface {} has no IP addresses, but expected network {} was configured.",
                                config.vpn_interface_name,
                                expected_network
                            ));
                            result.ip_network_match_status = IpNetworkMatchStatus::NoIPs(*expected_network);
                        } else {
                            for ip_net in &iface.ips {
                                if expected_network.contains(ip_net.ip()) {
                                    info!("Found expected IP {} on VPN interface {}.", ip_net, config.vpn_interface_name);
                                    found_matching_ip = true;
                                    break;
                                }
                            }
                            if found_matching_ip {
                                result.ip_network_match_status = IpNetworkMatchStatus::Match(*expected_network);
                            } else {
                                result.findings.push(format!(
                                    "No IP address on VPN interface {} belongs to the expected network {}. Found IPs: {:?}",
                                    config.vpn_interface_name,
                                    expected_network,
                                    iface.ips
                                ));
                                result.ip_network_match_status = IpNetworkMatchStatus::Mismatch(*expected_network);
                            }
                        }
                    }
                    None => {
                        // No expected network configured, just check if *any* IP exists.
                         if iface.ips.is_empty() {
                            result.findings.push(format!(
                                "Warning: VPN interface {} has no assigned IP addresses. (No expected network was configured).",
                                config.vpn_interface_name
                            ));
                             result.ip_network_match_status = IpNetworkMatchStatus::InterfaceHasNoIPs;
                        } else {
                            info!("IP addresses found for {}: {:?}. (No specific network expected).", config.vpn_interface_name, iface.ips);
                            // Status remains NotChecked as no expectation was set.
                        }
                    }
                }

                // Perform external IP check if enabled
                if config.check_external_ip.unwrap_or(false) {
                    info!("--- Performing External IP Check ---");
                    warn!("OPSEC WARNING: External IP check enabled. This sends traffic outside the VPN potentially, revealing real IP if VPN fails.");
                    match perform_external_ip_check(config) {
                        Ok(reported_ip) => {
                            info!("External IP check service reported IP: {}", reported_ip);
                            if let Some(expected_ips) = &config.expected_external_ips {
                                if expected_ips.is_empty() {
                                    result.findings.push("Warning: External IP check enabled, but 'expected_external_ips' list is empty. Cannot verify.".to_string());
                                    result.external_ip_status = ExternalIpStatus::CheckedOk(reported_ip); // OK, but can't verify
                                } else if expected_ips.contains(&reported_ip) {
                                    info!("Reported external IP matches one of the expected IPs.");
                                    result.external_ip_status = ExternalIpStatus::ExpectedMatch(reported_ip);
                                } else {
                                    let finding = format!(
                                        "Reported external IP {} does NOT match any expected IPs: {:?}",
                                        reported_ip, expected_ips
                                    );
                                    result.findings.push(finding);
                                    result.external_ip_status = ExternalIpStatus::ExpectedMismatch(reported_ip, expected_ips.clone());
                                }
                            } else {
                                // Check ran successfully, but no IPs were expected
                                result.external_ip_status = ExternalIpStatus::CheckedOk(reported_ip);
                            }
                        }
                        Err(e) => {
                             let finding = format!("External IP check failed: {}", e);
                             result.findings.push(finding.clone());
                             result.external_ip_status = ExternalIpStatus::CheckFailed(e);
                        }
                    }
                } else {
                    info!("--- Skipping External IP Check (disabled) ---");
                    result.external_ip_status = ExternalIpStatus::Disabled;
                }
            } // End if iface.is_up()
        }
        None => {
            result.findings.push(format!(
                "Configured VPN interface {} not found.", config.vpn_interface_name
            ));
            result.interface_found = false;
        }
    }

    // Log collected findings at the end
    for finding in &result.findings {
        // Use error! for definite problems, warn! for potential issues
        if finding.starts_with("Warning:") {
             warn!("{}", finding);
        } else {
             error!("{}", finding); // Treat other findings as errors for now
        }
    }

    result
}

/// Performs the external IP check using the configured URL.
/// (Function remains largely the same, just used by the refactored verify_network_interfaces)
///
/// # Arguments
/// * `config` - The application configuration.
/// # Returns
/// A `Result` containing the reported `IpAddr` or an error string.
fn perform_external_ip_check(config: &Config) -> Result<IpAddr, String> {
    let url = config.external_ip_check_url.as_deref().unwrap_or("https://ifconfig.me/ip");
    debug!("Querying external IP check service: {}", url);

    let client = Client::builder()
        .timeout(std::time::Duration::from_secs(10)) // Add a timeout
        .build()
        .map_err(|e| format!("Failed to build HTTP client: {}", e))?;

    let response = client.get(url)
        .send()
        .map_err(|e| format!("HTTP request to {} failed: {}", url, e))?;

    if !response.status().is_success() {
        return Err(format!("External IP check service {} returned status: {}", url, response.status()));
    }

    let content_type = response.headers()
        .get(reqwest::header::CONTENT_TYPE)
        .and_then(|val| val.to_str().ok())
        .unwrap_or("");

    let ip_str = if content_type.contains("application/json") {
        // Try parsing as JSON: {"ip": "..."}
        let json_response: IpCheckResponse = response.json()
            .map_err(|e| format!("Failed to parse JSON response from {}: {}", url, e))?;
        json_response.ip
    } else {
        // Assume plain text IP
        response.text()
            .map_err(|e| format!("Failed to read text response from {}: {}", url, e))?
            .trim().to_string()
    };

    // Parse the extracted string as an IP address
    ip_str.parse::<IpAddr>()
        .map_err(|e| format!("Invalid IP address format received from {}: '{}' ({})", url, ip_str, e))
}

/// Verifies DNS configuration by reading /etc/resolv.conf and comparing against expected servers.
/// Collects findings instead of returning early errors.
///
/// # Arguments
/// * `config` - The application configuration (hardcoded defaults).
/// # Returns
/// A `DnsCheckResult` struct containing detailed results and findings.
pub fn verify_dns(config: &Config) -> DnsCheckResult {
    info!("Verifying DNS configuration...");

    let expected_servers = config.expected_dns_servers.clone().unwrap_or_default();

    let mut result = DnsCheckResult {
        servers_expected: expected_servers.clone(),
        servers_found: Vec::new(),
        match_status: DnsMatchStatus::NotChecked, // Default
        findings: Vec::new(),
    };

    if expected_servers.is_empty() {
        warn!("No expected DNS servers configured. Skipping DNS check.");
        // Status remains NotChecked
        return result;
    }

    let resolv_conf_path = "/etc/resolv.conf";
    debug!("Reading DNS configuration from {}", resolv_conf_path);

    let content = match fs::read_to_string(resolv_conf_path) {
        Ok(c) => c,
        Err(e) => {
            let finding = format!("Failed to read {}: {}", resolv_conf_path, e);
            result.findings.push(finding.clone());
            result.match_status = DnsMatchStatus::ReadError(e.to_string());
            error!("{}", finding);
            return result; // Cannot proceed without the file content
        }
    };

    // Regex to find lines starting with "nameserver" followed by an IP address.
    let re = match Regex::new(r"^\s*nameserver\s+([^\s]+)") {
        Ok(r) => r,
        Err(e) => {
            // This is a programming error (invalid regex), should likely panic or log severe error.
            let finding = format!("FATAL: Failed to compile regex for DNS parsing: {}", e);
            result.findings.push(finding.clone());
            result.match_status = DnsMatchStatus::ReadError("Regex compilation failed".to_string());
            error!("{}", finding);
            // Consider returning here as parsing won't work
            return result;
        }
    };

    for line in content.lines() {
        let trimmed_line = line.trim();
        if trimmed_line.is_empty() || trimmed_line.starts_with('#') {
            continue;
        }

        if let Some(caps) = re.captures(trimmed_line) {
            if let Some(server_str_match) = caps.get(1) {
                let server_str = server_str_match.as_str();
                match server_str.parse::<IpAddr>() {
                    Ok(ip_addr) => result.servers_found.push(ip_addr),
                    Err(e) => {
                        // Log a warning but continue parsing other lines
                        warn!(
                            "Failed to parse potential nameserver IP '{}' from {}: {}. Skipping.",
                            server_str, resolv_conf_path, e
                        );
                        // Optionally add to findings if strict parsing is desired
                        // result.findings.push(format!("Warning: Unparseable nameserver entry '{}'", server_str));
                    }
                }
            }
        }
    }

    if result.servers_found.is_empty() {
         let finding = format!("No valid nameserver entries found in {}", resolv_conf_path);
         result.findings.push(finding.clone());
         result.match_status = DnsMatchStatus::NoServersFound;
         error!("{}", finding);
         return result; // No servers found to compare
    }

    info!("DNS servers found in {}: {:?}", resolv_conf_path, result.servers_found);
    debug!("Expected DNS servers: {:?}", result.servers_expected);

    // Compare the found servers with the expected servers using HashSets for order-insensitivity.
    let actual_set: HashSet<_> = result.servers_found.iter().collect();
    let expected_set: HashSet<_> = result.servers_expected.iter().collect();

    if actual_set == expected_set {
        info!("Actual DNS servers match expected configuration.");
        result.match_status = DnsMatchStatus::Match;
    } else {
        let missing: Vec<_> = expected_set.difference(&actual_set).map(|&&ip| ip.to_string()).collect();
        let unexpected: Vec<_> = actual_set.difference(&expected_set).map(|&&ip| ip.to_string()).collect();

        let mut error_msg = String::from("DNS server configuration mismatch.");
        if !missing.is_empty() {
            error_msg.push_str(&format!(" Missing expected servers: {:?}", missing));
        }
        if !unexpected.is_empty() {
            error_msg.push_str(&format!(" Found unexpected servers: {:?}", unexpected));
        }
        result.findings.push(error_msg.clone());
        result.match_status = DnsMatchStatus::Mismatch;
        error!("{}", error_msg); // Log the mismatch as an error
    }

    result
} 